cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -4927,6 +4927,160 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
                 }
             }
         }
+        Opcode::IaddPairwise => {
+            if let Some(swiden_low) = matches_input(ctx, inputs[0], Opcode::SwidenLow) {
+                if let Some(swiden_high) = matches_input(ctx, inputs[1], Opcode::SwidenHigh) {
+                    let swiden_input = &[
+                        InsnInput {
+                            insn: swiden_low,
+                            input: 0,
+                        },
+                        InsnInput {
+                            insn: swiden_high,
+                            input: 0,
+                        },
+                    ];
+
+                    let input0_ty = ctx.input_ty(swiden_low, 0);
+                    let input1_ty = ctx.input_ty(swiden_high, 0);
+                    let output_ty = ctx.output_ty(insn, 0);
+                    let src = put_input_in_reg(ctx, swiden_input[0]);
+                    let dst = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
+
+                    assert_eq!(input0_ty, input1_ty);
+
+                    match (input0_ty, output_ty) {
+                        (types::I8X16, types::I16X8) => {
+                            static MUL_CONST: [u8; 16] = [0x01; 16];
+                            let mul_const =
+                                ctx.use_constant(VCodeConstantData::WellKnown(&MUL_CONST));
+                            let mul_const_reg = ctx.alloc_tmp(types::I8X16).only_reg().unwrap();
+                            ctx.emit(Inst::xmm_load_const(mul_const, mul_const_reg, types::I8X16));
+                            ctx.emit(Inst::xmm_mov(
+                                SseOpcode::Movdqa,
+                                RegMem::reg(mul_const_reg.to_reg()),
+                                dst,
+                            ));
+                            ctx.emit(Inst::xmm_rm_r(SseOpcode::Pmaddubsw, RegMem::reg(src), dst));
```

Comment:
Sorry, I don't know anything about the ISA... but since you're using 'src' here, shouldn't you be checking that the input to the swiden_low and swiden_high is the same?

Reply:
+1, this needs to be part of the condition above (where we check input opcodes) such that we fall back to generic codegen otherwise, as it's perfectly legal to pairwise-add extends of two different values.

Reply:
@sparker-arm @cfallin .. This is exactly what I was thinking about. This is the tension between treating a clif as a 1:1 wasm instruction mapping versus the decomposition approach. I didn't add the check because as noted, it is perfect legal for swiden_low and swiden_high to have to not be the same per the definition of iadd_pairwise. Here, we are not lowering ext_addpairwise which would require the condition that the input be the same ... we are lowering iadd_pairwise. I was concerned that some exercise or fuzz testing or something lowering just this cliff instruction could be invalidated incorrectly. I can add the check as I want this to merge, but I do think we need to think about or even audit the consistency of similar checks and whether they map to the clif instruction that we are actually lowering or the wasm instruction that we are intending to lower.

Reply:
> Here, we are not lowering ext_addpairwise which would require the condition that the input be the same ... we are lowering iadd_pairwise

Well, that's true in the outer scope, but the instructions that this case is emitting are specifically for the same-input case, no?

IMHO it's fine for now to fall back to `unimplemented!()` if we know for sure that we won't generate an `iadd_pairwise` with other (arbitrary) inputs, but at some point we should fill in that case too, for completeness.

> should audit ...

I agree! I have lots of thoughts on verification that this text box is too small to contain :-) IMHO this is the next big effort after getting current projects (including some sort of isel DSL) landed.